llvm/labs/lab2/kuznetsov_mikhail/MGKFuncWrapper.cpp

```diff
@@ -0,0 +1,82 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+struct FunctionInstrumentation : llvm::PassInfoMixin<FunctionInstrumentation> {
+  llvm::PreservedAnalyses run(llvm::Function &func,
+                              llvm::FunctionAnalysisManager &) {
+    llvm::LLVMContext &ctx = func.getContext();
+    llvm::IRBuilder<> builder(ctx);
+    llvm::Module *mod = func.getParent();
+    bool hasStart = false;
+    bool hasEnd = false;
+
+    llvm::FunctionType *funcType =
+        llvm::FunctionType::get(llvm::Type::getVoidTy(ctx), false);
+    llvm::FunctionCallee startInstr =
+        mod->getOrInsertFunction("start_instrument", funcType);
+    llvm::FunctionCallee endInstr =
+        mod->getOrInsertFunction("end_instrument", funcType);
+
+    llvm::Function *startFunc = llvm::dyn_cast<llvm::Function>(
```

Comment:
dyn_cast may return nullptr. So, it is dangerous to dereference pointer without a check. But if you are sure that there is a function object always, you may use llvm::cast instead of dyn_cast.

Reply:
I meant that if you created a function object here then you could just cast it without an explicit check of type but your option is also fine.